##################################################################
Date: Tue Feb 18 17:12:03 2020 +0000
From:
Subject: [PATCH] Create storage directory before starting service
Source: SKY
Upstream-Status: Pending
Signed-off-by: Tijo Thomas John <tijo.thomas@sky.uk>
##################################################################

Index: bluez-5.82/src/bluetooth.service.in
===================================================================
--- bluez-5.82.orig/src/bluetooth.service.in
+++ bluez-5.82/src/bluetooth.service.in
@@ -8,6 +8,15 @@ ConditionPathIsDirectory=/sys/class/blue
 [Service]
 Type=dbus
 BusName=org.bluez
+
+ExecStartPre=/bin/mkdir -p /var/lib/bluetooth
+ExecStartPre=/bin/mkdir -p /opt/lib/bluetooth
+
+ExecStartPre=/bin/sh -c 'if [ -d /opt/secure/lib/bluetooth ]; then \
+    cp -r /opt/secure/lib/bluetooth /opt/lib/ && \
+    rm -r /opt/secure/lib/bluetooth; \
+fi'
+
 ExecStart=/usr/lib/bluez5/bluetooth/bluetoothd -d
 NotifyAccess=main
 #WatchdogSec=10
