From b0343c46ecff5e5b9ebb7638ef36d300205ad235 Mon Sep 17 00:00:00 2001
From: Gurdal Oruklu <gurdal_oruklu@comcast.com>
Date: Thu, 19 Jan 2023 15:12:57 +0000
Subject: [PATCH] set brcmpcmsink low_latency property for webrtc streams

Signed-off-by: Gurdal Oruklu <gurdal_oruklu@comcast.com>
---
 .../graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp   | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index 0c4fb5c88352..cf5231300c0a 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -3902,6 +3902,15 @@ void MediaPlayerPrivateGStreamer::configureElement(GstElement* element)
             g_object_set(G_OBJECT(element), "limit_buffering_ms", 1000, nullptr);
         }
     }
+    if(!g_strcmp0(G_OBJECT_TYPE_NAME(G_OBJECT(element)), "GstBrcmPCMSink")) {
+#if ENABLE(MEDIA_STREAM)
+        if (m_streamPrivate != nullptr && g_object_class_find_property(G_OBJECT_GET_CLASS(element), "low_latency")) {
+            GST_DEBUG("Set 'low_latency' in brcmpcmsink");
+            g_object_set (G_OBJECT(element), "low_latency", TRUE, nullptr);
+            g_object_set (G_OBJECT(element), "low_latency_max_queued_ms", 60, nullptr);
+        }
+#endif
+    }
 #endif
 
 #if USE(WESTEROS_SINK)
-- 
2.17.1

