From b1bc9c60787185d56a2e60772ddfb42fc7d26364 Mon Sep 17 00:00:00 2001
From: krishna priya kanagaraj <krishnapriya_kanagaraj@comcast.com>
Date: Tue, 7 Feb 2023 08:33:03 +0000
Subject: [PATCH] BCOM-6652-fix-gst1_18-playbin3-autoplug

---
 .../graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp        | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index 2ee80742fbcf..5c0afba3bcbc 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -3132,7 +3132,11 @@ void MediaPlayerPrivateGStreamer::elementSetupCallback(MediaPlayerPrivateGStream
 #endif
     }
     // FIXME: Following is a hack needed to get westeros-sink autoplug correctly with playbin3.
+#if GST_CHECK_VERSION(1, 18, 0)
+    if (!player->m_isLegacyPlaybin && westerosSinkCaps && g_str_has_prefix(GST_ELEMENT_NAME(element), "uridecodebin3")) {
+#else
     if (!player->m_isLegacyPlaybin && westerosSinkCaps && g_str_has_prefix(GST_ELEMENT_NAME(element), "decodebin3")) {
+#endif
         GstCaps* defaultCaps = nullptr;
         g_object_get (element, "caps", &defaultCaps, NULL);
         defaultCaps = gst_caps_merge(defaultCaps, gst_caps_ref(westerosSinkCaps));
-- 
2.17.1


