From 05307a92a0030860d798b20f83ca436ba3a2e921 Mon Sep 17 00:00:00 2001
From: Gurdal Oruklu <gurdal_oruklu@comcast.com>
Date: Tue, 31 Jan 2023 03:56:40 +0000
Subject: [PATCH] set brcmpcmsink low_latency property for webrtc streams

Signed-off-by: Gurdal Oruklu <gurdal_oruklu@comcast.com>
---
 .../graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp    | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
index 20cfe9411b87..a1d608dea4a8 100644
--- a/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp
@@ -3098,6 +3098,14 @@ void MediaPlayerPrivateGStreamer::elementSetupCallback(MediaPlayerPrivateGStream
     }
     else if (g_str_has_prefix(GST_ELEMENT_NAME(element), "brcmaudiodecoder")) {
         g_object_set(G_OBJECT(element), "audio_pts_disco_threshold", G_MAXUINT, nullptr);
+    }else if(!g_strcmp0(G_OBJECT_TYPE_NAME(G_OBJECT(element)), "GstBrcmPCMSink")) {
+#if ENABLE(MEDIA_STREAM)
+        if (player->m_streamPrivate != nullptr && g_object_class_find_property(G_OBJECT_GET_CLASS(element), "low_latency")) {
+            GST_DEBUG("Set 'low_latency' in brcmpcmsink");
+            g_object_set (G_OBJECT(element), "low_latency", TRUE, nullptr);
+            g_object_set (G_OBJECT(element), "low_latency_max_queued_ms", 60, nullptr);
+        }
+#endif
     }
 #endif
 
-- 
2.17.1

