From 9f9cb309e3c109b887e9b1a15898b724fd6acec3 Mon Sep 17 00:00:00 2001
From: Michael Pantazoglou <michael.pantazoglou@oceanbluesoftware.co.uk>
Date: Tue, 11 Apr 2023 17:50:10 +0300
Subject: [PATCH] Create XML document for HbbTV XHTML MIME type

---
 Source/WebCore/dom/DOMImplementation.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Source/WebCore/dom/DOMImplementation.cpp b/Source/WebCore/dom/DOMImplementation.cpp
index 9edc20da78f02..80c2d68f98204 100644
--- a/Source/WebCore/dom/DOMImplementation.cpp
+++ b/Source/WebCore/dom/DOMImplementation.cpp
@@ -143,7 +143,7 @@ Ref<Document> DOMImplementation::createDocument(const String& type, Frame* frame
     // Plug-ins cannot take over for HTML, XHTML, plain text, or non-PDF images.
     if (equalLettersIgnoringASCIICase(type, "text/html"))
         return HTMLDocument::create(frame, url);
-    if (equalLettersIgnoringASCIICase(type, "application/xhtml+xml"))
+    if (equalLettersIgnoringASCIICase(type, "application/xhtml+xml") || equalLettersIgnoringASCIICase(type, "application/vnd.hbbtv.xhtml+xml"))
         return XMLDocument::createXHTML(frame, url);
     if (equalLettersIgnoringASCIICase(type, "text/plain"))
         return TextDocument::create(frame, url);
