Index: bluez-5.75/src/bluetooth.service.in
===================================================================
--- bluez-5.75.orig/src/bluetooth.service.in
+++ bluez-5.75/src/bluetooth.service.in
@@ -8,7 +8,7 @@ ConditionPathIsDirectory=/sys/class/blue
 [Service]
 Type=dbus
 BusName=org.bluez
-ExecStart=@PKGLIBEXECDIR@/bluetoothd
+ExecStart=/usr/lib/bluez5/bluetooth/bluetoothd -d
 NotifyAccess=main
 #WatchdogSec=10
 Restart=always
