Index: bluez-5.77/src/adapter.c
===================================================================
--- bluez-5.77.orig/src/adapter.c
+++ bluez-5.77/src/adapter.c
@@ -24,6 +24,7 @@
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <dirent.h>
+#include <fcntl.h>
 
 #include <glib.h>
 #include <dbus/dbus.h>
@@ -172,6 +173,8 @@ static GSList *adapter_drivers = NULL;
 static GSList *disconnect_list = NULL;
 static GSList *conn_fail_list = NULL;
 
+extern struct hid_firm_info *head;
+
 struct link_key_info {
 	bdaddr_t bdaddr;
 	uint8_t bdaddr_type;
@@ -4293,6 +4296,30 @@ static int set_privacy(struct btd_adapte
 	return -1;
 }
 
+static void set_adapter_privacy(struct btd_adapter *adapter, uint8_t privacy)
+{
+        struct mgmt_cp_set_privacy cp;
+        int fd;
+
+        fd = open("/dev/urandom", O_RDONLY);
+        if (fd < 0) {
+                DBG("Not able to open the file");
+                return -1;
+        }
+
+        if (read(fd, cp.irk, sizeof(cp.irk)) != sizeof(cp.irk)) {
+                DBG("Failed to read the irk key");
+                close(fd);
+                return -1;
+        }
+
+        cp.privacy = privacy;
+        close(fd);
+
+        if (mgmt_send(adapter->mgmt, MGMT_OP_SET_PRIVACY,adapter->dev_id, sizeof(cp), &cp, set_privacy_complete, adapter, NULL) > 0)
+                return 0;
+}
+
 static void load_link_keys(struct btd_adapter *adapter, bool debug_keys,
 							bool retry);
 
@@ -10898,6 +10925,15 @@ void adapter_shutdown(void)
 		adapter_remaining++;
 	}
 
+        struct hid_firm_info *current = head;
+        struct hid_firm_info *next;
+        while (current != NULL) {
+                next = current->next;
+                free(current);
+                current = next;
+        }
+        head = NULL;
+
 	if (!adapter_remaining)
 		btd_exit();
 }
@@ -10942,3 +10978,21 @@ bool btd_adapter_has_exp_feature(struct
 
 	return false;
 }
+
+void change_adapter_privacy_mode(struct btd_adapter *adapter,uint8_t privacy)
+{
+        DBG("Setting privacy mode : %u",privacy);
+        set_adapter_privacy(adapter,privacy);
+        btd_opts.privacy = privacy;
+}
+
+bool get_privacy_status (struct btd_adapter *adapter)
+{
+        if (btd_opts.privacy == 0x00) {
+                DBG("Privacy Disabled");
+                return false;
+        } else if (btd_opts.privacy == 0x01) {
+                DBG("Privacy enabled");
+                return true;
+        }
+}
Index: bluez-5.77/src/adapter.h
===================================================================
--- bluez-5.77.orig/src/adapter.h
+++ bluez-5.77/src/adapter.h
@@ -298,3 +298,7 @@ void btd_adapter_store_conn_param(struct
 				const bdaddr_t *peer, uint8_t bdaddr_type,
 				uint16_t min_interval, uint16_t max_interval,
 				uint16_t latency, uint16_t timeout);
+
+bool get_privacy_status (struct btd_adapter *adapter);
+
+void change_adapter_privacy_mode(struct btd_adapter *adapter,uint8_t privacy);
Index: bluez-5.77/src/device.c
===================================================================
--- bluez-5.77.orig/src/device.c
+++ bluez-5.77/src/device.c
@@ -86,6 +86,8 @@
 static DBusConnection *dbus_conn = NULL;
 static unsigned service_state_cb_id;
 
+struct hid_firm_info *head = NULL;
+
 struct btd_disconnect_data {
 	guint id;
 	disconnect_watch watch;
@@ -246,6 +248,7 @@ struct btd_device {
 	unsigned int	disconn_timer;
 	unsigned int	discov_timer;
 	unsigned int	temporary_timer;	/* Temporary/disappear timer */
+	unsigned int	privacy_timer;
 	struct browse_req *browse;		/* service discover request */
 	struct bonding_req *bonding;
 	struct authentication_req *authr;	/* authentication request */
@@ -567,11 +570,106 @@ static bool device_address_is_private(st
 	}
 }
 
+static void device_set_auto_connect(struct btd_device *device, gboolean enable)
+{
+        char addr[MAX_ADDR_SIZE];
+
+        if (!device || !device->le)
+                return;
+
+        ba2str(&device->bdaddr, addr);
+
+        DBG("%s auto connect: %d", addr, enable);
+
+        if (device->auto_connect == enable)
+                return;
+
+        device->auto_connect = enable;
+
+        /* Disabling auto connect */
+        if (enable == FALSE) {
+                adapter_connect_list_remove(device->adapter, device);
+                adapter_auto_connect_remove(device->adapter, device);
+                return;
+        }
+
+        /* Enabling auto connect */
+        adapter_auto_connect_add(device->adapter, device);
+
+        if (device->attrib) {
+                DBG("Already connected");
+                return;
+        }
+
+        adapter_connect_list_add(device->adapter, device);
+}
+
+int store_hid_info(char *modalias,char *addr)
+{
+        struct hid_firm_info *data = NULL,*temp = NULL;
+
+        DBG("");
+        data = g_new0(struct hid_firm_info, sizeof(struct hid_firm_info));
+        strncpy (data->hid_mac_address, addr, MAX_ADDR_SIZE);
+        strncpy (data->hid_firm_version, modalias, MAX_FIRM_VER_SIZE);
+
+        if (head == NULL) {
+                data->next = head;
+                head = data;
+        } else {
+                temp = head;
+                while (temp->next != NULL)
+                        temp = temp->next;
+                data->next = temp->next;
+                temp->next = data;
+        }
+        temp = NULL;
+        return 0;
+}
+
+bool get_hid_info(char *modalias,char *addr)
+{
+        struct hid_firm_info *current = head;
+        while (current != NULL) {
+                if ((strncmp(current->hid_firm_version, modalias, MAX_FIRM_VER_SIZE - 1) == 0) && (strncmp(current->hid_mac_address, addr, MAX_ADDR_SIZE - 1) == 0)) {
+                        return TRUE;
+                }
+                current = current->next;
+        }
+        return FALSE;
+}
+
 static void store_device_info(struct btd_device *device)
 {
+        bool info_available = FALSE;
+        char addr[MAX_ADDR_SIZE];
+        int ret;
+        gboolean privacy_status = FALSE;
+
 	if (device->temporary || device->store_id > 0)
 		return;
 
+        ba2str(&device->bdaddr, addr);
+        struct bearer_state *state = get_state(device, device->bdaddr_type);
+        DBG("con - %d pair - %d bon - %d ",state->connected,state->paired,state->bonded);
+
+        if (g_strcmp0(device->modalias,HID_XBOX_GEN4_DEFAULT_FIRM) == 0) {
+                DBG("Checking the stored hid info for version - %s addr - %s",device->modalias,addr);
+                info_available = get_hid_info(device->modalias,addr);
+                if (info_available == TRUE) {
+                        DBG("Device already present ,Disabling the privacy");
+                } else {
+                        DBG("Storing the device details ..");
+                        ret = store_hid_info(device->modalias,addr);
+                        if (ret == 0) {
+                                DBG("HID details stored");
+                        } else {
+                                DBG("Storing device details failed");
+                        }
+                        DBG("New xbox gen4 found with default firmware,Re-pairing required with privacy enabled %d ",device->le_state.connected);
+                }
+        }
+
 	if (device_address_is_private(device)) {
 		DBG("Can't store info for private addressed device %s",
 								device->path);
@@ -2065,40 +2163,6 @@ bool btd_device_add_set(struct btd_devic
 	return true;
 }
 
-static void device_set_auto_connect(struct btd_device *device, gboolean enable)
-{
-	char addr[18];
-
-	if (!device || !device->le || device_address_is_private(device))
-		return;
-
-	ba2str(&device->bdaddr, addr);
-
-	DBG("%s auto connect: %d", addr, enable);
-
-	if (device->auto_connect == enable)
-		return;
-
-	device->auto_connect = enable;
-
-	/* Disabling auto connect */
-	if (enable == FALSE) {
-		adapter_connect_list_remove(device->adapter, device);
-		adapter_auto_connect_remove(device->adapter, device);
-		return;
-	}
-
-	/* Enabling auto connect */
-	adapter_auto_connect_add(device->adapter, device);
-
-	if (device->attrib) {
-		DBG("Already connected");
-		return;
-	}
-
-	adapter_connect_list_add(device->adapter, device);
-}
-
 static DBusMessage *dev_disconnect(DBusConnection *conn, DBusMessage *msg,
 							void *user_data)
 {
@@ -2573,11 +2637,25 @@ static uint8_t select_conn_bearer(struct
 	return dev->bdaddr_type;
 }
 
+static gboolean disable_privacy_feature(gpointer user_data)
+{
+       struct btd_device *device = user_data;
+       DBG("In disable privacy callback");
+       change_adapter_privacy_mode(device->adapter,0x00);
+       device->disable_auto_connect = TRUE;
+       device_set_auto_connect(device, FALSE);
+       g_source_remove(device->privacy_timer);
+       return FALSE;
+}
+
 static DBusMessage *dev_connect(DBusConnection *conn, DBusMessage *msg,
 							void *user_data)
 {
 	struct btd_device *dev = user_data;
 	uint8_t bdaddr_type;
+        char addr[18];
+        struct btd_adapter *adapter = dev->adapter;
+        gboolean privacy_status = false;
 
 	if (dev->bredr_state.connected) {
 		/*
@@ -2612,9 +2690,19 @@ static DBusMessage *dev_connect(DBusConn
 		if (dev->le_state.connected)
 			return dbus_message_new_method_return(msg);
 
+                privacy_status = get_privacy_status(adapter);
+                ba2str(&dev->bdaddr, addr);
+                struct bearer_state *state = get_state(dev, bdaddr_type);
+                if ((get_hid_info(HID_XBOX_GEN4_DEFAULT_FIRM,addr) == TRUE) && privacy_status == FALSE && state->paired) {
+                        DBG(" Connection initiated for xbox gen4 with default firmware ,Privacy need to be enabled ");
+                        change_adapter_privacy_mode(adapter,0x01);
+                        dev->privacy_timer = g_timeout_add_seconds(ADAPTER_PAIR_PRIVACY_TIMEOUT,disable_privacy_feature,dev);
+                }
+
 		btd_device_set_temporary(dev, false);
 
-		if (dev->disable_auto_connect) {
+		if (dev->disable_auto_connect && (get_hid_info(HID_XBOX_GEN4_DEFAULT_FIRM,addr) != TRUE)) {
+                        DBG("Setting auto connect ");
 			dev->disable_auto_connect = FALSE;
 			device_set_auto_connect(dev, TRUE);
 		}
@@ -3084,6 +3172,8 @@ static DBusMessage *pair_device(DBusConn
 	struct bonding_req *bonding;
 	uint8_t io_cap;
 	int err;
+        char addr[18];
+        gboolean privacy_status = false;
 
 	btd_device_set_temporary(device, false);
 
@@ -3093,6 +3183,14 @@ static DBusMessage *pair_device(DBusConn
 	if (device->bonding)
 		return btd_error_in_progress(msg);
 
+        privacy_status = get_privacy_status(adapter);
+        ba2str(&device->bdaddr, addr);
+        if ((get_hid_info(HID_XBOX_GEN4_DEFAULT_FIRM,addr) == TRUE) && device->appearance == 0x3c4 && privacy_status == FALSE) {
+                DBG("Pairing initiated for xbox gen4 with default firmware ,Privacy need to be enabled ");
+                change_adapter_privacy_mode(adapter,0x01);
+                device->privacy_timer = g_timeout_add_seconds(ADAPTER_CONNECT_PRIVACY_TIMEOUT,disable_privacy_feature,device);
+        }
+
 	if (device->bredr_state.bonded)
 		bdaddr_type = device->bdaddr_type;
 	else if (device->le_state.bonded)
Index: bluez-5.77/src/device.h
===================================================================
--- bluez-5.77.orig/src/device.h
+++ bluez-5.77/src/device.h
@@ -10,6 +10,20 @@
  */
 
 #define DEVICE_INTERFACE	"org.bluez.Device1"
+#define MAX_FIRM_VER_SIZE 20
+#define MAX_ADDR_SIZE 18
+#define ADAPTER_PAIR_PRIVACY_TIMEOUT 4
+#define ADAPTER_CONNECT_PRIVACY_TIMEOUT 2
+#define HID_XBOX_GEN4_DEFAULT_FIRM "usb:v045Ep0B13d0509"
+
+struct hid_firm_info {
+        char hid_mac_address[MAX_ADDR_SIZE];
+        char hid_firm_version[MAX_FIRM_VER_SIZE];
+        struct hid_firm_info *next;
+};
+
+bool get_hid_info(char *modalias,char *addr);
+int store_hid_info(char *modalias,char *addr);
 
 struct btd_device;
 
