Index: git/fscryptctl.c
===================================================================
--- git.orig/fscryptctl.c
+++ git/fscryptctl.c
@@ -91,6 +91,8 @@ enum {
   OPT_PADDING,
 };
 
+char identifier_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
+
 /* util-linux style usage */
 static void __attribute__((__noreturn__)) usage(FILE *out) {
   fputs(
@@ -375,7 +377,7 @@ static bool set_policy(const char *path,
 //                                 Commands
 // -----------------------------------------------------------------------------
 
-static int cmd_add_key(int argc, char *const argv[]) {
+int cmd_add_key(int argc, char *const argv[]) {
   handle_no_options(&argc, &argv);
   if (argc != 1) {
     fputs("error: must specify a single mountpoint\n", stderr);
@@ -410,7 +412,6 @@ static int cmd_add_key(int argc, char *c
   }
   close(fd);
 
-  char identifier_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
   bytes_to_hex(arg->key_spec.u.identifier, FSCRYPT_KEY_IDENTIFIER_SIZE,
                identifier_hex);
   puts(identifier_hex);
@@ -421,7 +422,7 @@ cleanup:
   return status;
 }
 
-static int cmd_remove_key(int argc, char *const argv[]) {
+int cmd_remove_key(int argc, char *const argv[]) {
   unsigned long ioc = FS_IOC_REMOVE_ENCRYPTION_KEY;
 
   static const struct option remove_key_options[] = {
@@ -473,7 +474,7 @@ static int cmd_remove_key(int argc, char
   return EXIT_SUCCESS;
 }
 
-static int cmd_key_status(int argc, char *const argv[]) {
+int cmd_key_status(int argc, char *const argv[]) {
   handle_no_options(&argc, &argv);
   if (argc != 2) {
     fputs("error: must specify a key identifier and a mountpoint\n", stderr);
@@ -576,7 +577,6 @@ static void show_v1_encryption_policy(co
 }
 
 static void show_v2_encryption_policy(const struct fscrypt_policy_v2 *policy) {
-  char identifier_hex[FSCRYPT_KEY_IDENTIFIER_HEX_SIZE];
   bytes_to_hex(policy->master_key_identifier, FSCRYPT_KEY_IDENTIFIER_SIZE,
                identifier_hex);
   printf("\tMaster key identifier: %s\n", identifier_hex);
@@ -587,7 +587,7 @@ static void show_v2_encryption_policy(co
 
 // For a specified file or directory with encryption enabled, print the
 // corresponding policy to stdout.
-static int cmd_get_policy(int argc, char *const argv[]) {
+int cmd_get_policy(int argc, char *const argv[]) {
   handle_no_options(&argc, &argv);
   if (argc != 1) {
     fputs("error: must specify a single file or directory\n", stderr);
@@ -618,7 +618,7 @@ static int cmd_get_policy(int argc, char
 
 // Apply an encryption policy to the specified directory.  The encryption
 // options can be overridden by command-line options.
-static int cmd_set_policy(int argc, char *const argv[]) {
+int cmd_set_policy(int argc, char *const argv[]) {
   uint8_t contents_encryption_mode = FSCRYPT_MODE_AES_256_XTS;
   uint8_t filenames_encryption_mode = FSCRYPT_MODE_AES_256_CTS;
   // Default to maximum zero-padding to leak less info about filename lengths.
Index: git/Makefile
===================================================================
--- git.orig/Makefile
+++ git/Makefile
@@ -32,7 +32,7 @@ BINDIR ?= $(PREFIX)/bin
 MANDIR ?= $(PREFIX)/share/man
 
 # C compiler flags
-CFLAGS ?= -O2 -Wall
+CFLAGS ?= -O2 -Wall -fPIC
 
 # C preprocessor flags
 CPPFLAGS ?=
@@ -58,6 +58,8 @@ fscryptctl: $(OBJ)
 $(OBJ): %.o: %.c $(HDRS)
 	$(CC) -o $@ -c $(CPPFLAGS) $(CFLAGS) $<
 
+libfscryptctl.so: $(OBJ)
+	$(CC) -shared -o $@ $^
 ##############################################################################
 
 # Build the manual page
@@ -133,7 +135,7 @@ test-all:
 
 # Installation, uninstallation, and cleanup targets
 
-all:fscryptctl #fscryptctl.1
+all:fscryptctl libfscryptctl.so #fscryptctl.1
 
 .PHONY: all install install-bin install-man uninstall clean
 
@@ -152,7 +154,7 @@ uninstall:
 #	rm -f $(DESTDIR)$(MANDIR)/man1/fscryptctl.1
 
 clean:
-	rm -f fscryptctl fscryptctl.1 *.o *.pyc
+	rm -f fscryptctl fscryptctl.1 *.o *.pyc *.so
 	rm -rf __pycache__
 	rm -rf .pytest_cache
 
